### PR TITLE
Monix rc1

### PR DESCRIPTION
**Pull request by alari**

### CODE REVIEW

File: b-tree/client/src/main/scala/fluence/btree/client/MerkleBTreeClient.scala [hunk 1]
```diff
@@ -122,60 +121,62 @@ class MerkleBTreeClient[K] private (
...
-            _ ← putDetailsMVar.put(Some(putDetails))
-            _ ← merklePathMVar.put(mPath.add(nodeProofWithIdx))
-          } yield putDetails
-
-        } else {
-          Task.raiseError(
-            new IllegalStateException(
-              s"Checksum of leaf didn't pass verifying for key=$key, Leaf(${keys.mkString(",")}, ${keys.mkString(",")})"
+    override def putDetails(keys: Array[Key], values: Array[Hash]): Task[ClientPutDetails] =
+      merklePathMVar.flatMap { mvar ⇒
```
Comment (DieMyst) [inline b-tree/client/src/main/scala/fluence/btree/client/MerkleBTreeClient.scala]: maybe put flatMap chain into for-comprehension?

File: kademlia/core/src/main/scala/fluence/kad/Kademlia.scala [hunk 2]
```diff
@@ -162,17 +96,174 @@ abstract class Kademlia[F[_], C](
...
+    pingExpiresIn: Duration,
+    checkNode: Node[C] ⇒ F[Boolean],
+    ownContactGetter: F[Node[C]],
+    kademliaRpc: C ⇒ KademliaRpc[F, C]
+  )(
+    implicit F: MonadError[F, Throwable],
+    P: Parallel[F, G],
+    BW: Bucket.WriteOps[F, C],
+    SW: Siblings.WriteOps[F, C]
+  ): Kademlia[F, C] = new Kademlia[F, C] {
```
Comment (DieMyst) [inline kademlia/core/src/main/scala/fluence/kad/Kademlia.scala]: Can it be better to move the implementation to another file?

File: kademlia/monix/js/src/main/scala/fluence/kad/MutableMapMVarBucketOps.scala [hunk 3]
```diff
@@ -27,17 +27,20 @@ import monix.eval.{MVar, Task}
...
   import RunOnMVar.runOnMVar
 
   override protected def run[T](bucketId: Int, mod: StateT[Task, Bucket[C], T]): Task[T] =
-    runOnMVar(
-      writeState.getOrElseUpdate(bucketId, MVar(read(bucketId))),
-      mod,
-      readState.update(bucketId, _: Bucket[C])
-    )
+    for {
+      ws <- writeState.getOrElseUpdate(bucketId, MVar(read(bucketId)).memoize)
```
Comment (DieMyst) [inline kademlia/monix/js/src/main/scala/fluence/kad/MutableMapMVarBucketOps.scala]: reformat?

File: kademlia/monix/jvm/src/main/scala/fluence/kad/MVarMapCache.scala [hunk 4]
```diff
@@ -29,57 +29,69 @@ import scala.collection.concurrent.TrieMap
...
     readState.get(key)
 
   def getOrAdd(key: K, value: V): Task[V] =
-    runOnMVar(
-      writeState.getOrElseUpdate(key, MVar(value)),
-      StateT.get[Task, V],
-      readState.update(key, _: V)
-    )
+    for {
+      s <- getMVar(key, value)
```
Comment (DieMyst) [inline kademlia/monix/jvm/src/main/scala/fluence/kad/MVarMapCache.scala]: reformat does not work

File: b-tree/client/src/main/scala/fluence/btree/client/MerkleBTreeClient.scala (hunk 1 above)
Comment (RicoGit) [reply b-tree/client/src/main/scala/fluence/btree/client/MerkleBTreeClient.scala]: it's ok for now, MVar.memoize it's temp decision, we will change it later

File: kademlia/core/src/main/scala/fluence/kad/Kademlia.scala (hunk 2 above)
Comment (alari) [reply kademlia/core/src/main/scala/fluence/kad/Kademlia.scala]: nope

File: kademlia/monix/js/src/main/scala/fluence/kad/MutableMapMVarBucketOps.scala (hunk 3 above)
Comment (alari) [reply kademlia/monix/js/src/main/scala/fluence/kad/MutableMapMVarBucketOps.scala]: You should take a look on scala-fmt for js and jvm paths, probably

**Pull request merged**